### ISSUE TITLE
[Good First Issue] [NNCF] Cover NNCF code with least coverage percentage with unit tests to at least 80%

### ISSUE THREAD

**Issue by vshampor** [openvinotoolkit/nncf#2496]
## Context

[NNCF](https://github.com/openvinotoolkit/nncf) uses `coverage` to run its precommit checks and currently has an admirable coverage rate of 90% across the repository. However, there are still some files that have <80% coverage, some - with no coverage at all. Some of these are located in the `experimental` folder and do not have a strict coverage requirement, some are special files such as `__init__.py` or `setup.py` which are not easily unit-tested, but most have actual production code inside and many of those have exactly 0.00% coverage. We need such files covered by precommit-scope unit tests in order not to let potential bugs get away.

## Task
1. Review the current per-file coverage percentage at Codecov (sort by coverage)
https://app.codecov.io/gh/openvinotoolkit/nncf?search=&displayType=list

2. For the following files in the list (click to see per-line coverage for each on Codecov), add unit tests (using `pytest` style) under the `tests` directory (in a subdirectory corresponding to the tested backend, e.g. a test for `nncf/torch/foo.py` would go to `tests/torch/test_foo.py`, for `nncf/common/bar.py` - to `tests/torch/test_bar.py`).

- [ ] [nncf/common/utils/decorators.py](https://app.codecov.io/gh/openvinotoolkit/nncf/blob/develop/nncf/common/utils/decorators.py)
- [ ] [nncf/common/utils/os.py](https://app.codecov.io/gh/openvinotoolkit/nncf/blob/develop/nncf/common/utils/os.py)
- [ ] [nncf/common/utils/patcher.py](https://app.codecov.io/gh/openvinotoolkit/nncf/blob/develop/nncf/common/utils/patcher.py)
- [ ] [nncf/common/utils/timer.py](https://app.codecov.io/gh/openvinotoolkit/nncf/blob/develop/nncf/common/utils/timer.py)
- [ ] [nncf/config/schema.py](https://app.codecov.io/gh/openvinotoolkit/nncf/blob/develop/nncf/config/schema.py)
- [ ] [nncf/openvino/quantization/quantize_model.py](https://app.codecov.io/gh/openvinotoolkit/nncf/blob/develop/nncf/openvino/quantization/quantize_model.py)
- [ ] [nncf/quantization/algorithms/accuracy_control/algorithm.py](https://app.codecov.io/gh/openvinotoolkit/nncf/blob/develop/nncf/quantization/algorithms/accuracy_control/algorithm.py)
- [ ] [nncf/quantization/algorithms/accuracy_control/evaluator.py](https://app.codecov.io/gh/openvinotoolkit/nncf/blob/develop/nncf/quantization/algorithms/accuracy_control/evaluator.py)
- [ ] [nncf/quantization/algorithms/accuracy_control/openvino_backend.py](https://app.codecov.io/gh/openvinotoolkit/nncf/blob/develop/nncf/quantization/algorithms/accuracy_control/openvino_backend.py)
- [ ] [nncf/quantization/algorithms/accuracy_control/ranker.py](https://app.codecov.io/gh/openvinotoolkit/nncf/blob/develop/nncf/quantization/algorithms/accuracy_control/ranker.py)
- [ ] [nncf/quantization/algorithms/accuracy_control/subset_selection.py](https://app.codecov.io/gh/openvinotoolkit/nncf/blob/develop/nncf/quantization/algorithms/accuracy_control/subset_selection.py)
- [ ] [nncf/quantization/algorithms/hyperparameter_tuner/algorithm.py](https://app.codecov.io/gh/openvinotoolkit/nncf/blob/develop/nncf/quantization/algorithms/hyperparameter_tuner/algorithm.py)
- [ ] [nncf/quantization/algorithms/hyperparameter_tuner/param_grid.py](https://app.codecov.io/gh/openvinotoolkit/nncf/blob/develop/nncf/quantization/algorithms/hyperparameter_tuner/param_grid.py)
- [ ] [nncf/quantization/algorithms/post_training/algorithm.py](https://app.codecov.io/gh/openvinotoolkit/nncf/blob/develop/nncf/quantization/algorithms/post_training/algorithm.py)
- [ ] [nncf/telemetry/wrapper.py](https://app.codecov.io/gh/openvinotoolkit/nncf/blob/develop/nncf/telemetry/wrapper.py)
- [ ] [nncf/tensorflow/helpers/model_creation.py](https://app.codecov.io/gh/openvinotoolkit/nncf/blob/develop/nncf/tensorflow/helpers/model_creation.py)
- [ ] [nncf/tensorflow/pruning/filter_pruning/functions.py](https://app.codecov.io/gh/openvinotoolkit/nncf/blob/develop/nncf/tensorflow/pruning/filter_pruning/functions.py)
- [ ] [nncf/tensorflow/tensor_statistics/statistics.py](https://app.codecov.io/gh/openvinotoolkit/nncf/blob/develop/nncf/tensorflow/tensor_statistics/statistics.py)
- [ ] [nncf/tensorflow/tf_internals.py](https://app.codecov.io/gh/openvinotoolkit/nncf/blob/develop/nncf/tensorflow/tf_internals.py)
- [ ] [nncf/tensorflow/utils/node.py](https://app.codecov.io/gh/openvinotoolkit/nncf/blob/develop/nncf/tensorflow/utils/node.py)
- [ ] [nncf/tensorflow/utils/scopes_handle.py](https://app.codecov.io/gh/openvinotoolkit/nncf/blob/develop/nncf/tensorflow/utils/scopes_handle.py)
- [ ] [nncf/torch/accuracy_aware_training/utils.py](https://app.codecov.io/gh/openvinotoolkit/nncf/blob/develop/nncf/torch/accuracy_aware_training/utils.py)
- [ ] [nncf/torch/binarization/binarize_functions.py](https://app.codecov.io/gh/openvinotoolkit/nncf/blob/develop/nncf/torch/binarization/binarize_functions.py)
- [ ] [nncf/torch/binarization/extensions.py](https://app.codecov.io/gh/openvinotoolkit/nncf/blob/develop/nncf/torch/binarization/extensions.py)
- [ ] [nncf/torch/pruning/export_utils.py](https://app.codecov.io/gh/openvinotoolkit/nncf/blob/develop/nncf/torch/pruning/export_utils.py)
- [ ] [nncf/torch/pruning/operations.py](https://app.codecov.io/gh/openvinotoolkit/nncf/blob/develop/nncf/torch/pruning/operations.py)
- [ ] [nncf/torch/quantization/base_ctrl.py](https://app.codecov.io/gh/openvinotoolkit/nncf/blob/develop/nncf/torch/quantization/base_ctrl.py)
- [ ] [nncf/torch/quantization/precision_init/autoq_init.py](https://app.codecov.io/gh/openvinotoolkit/nncf/blob/develop/nncf/torch/quantization/precision_init/autoq_init.py)
- [ ] [nncf/torch/quantization/statistics.py](https://app.codecov.io/gh/openvinotoolkit/nncf/blob/develop/nncf/torch/quantization/statistics.py)
- [ ] [nncf/torch/sparsity/const/algo.py](https://app.codecov.io/gh/openvinotoolkit/nncf/blob/develop/nncf/torch/sparsity/const/algo.py)
- [ ] [nncf/torch/sparsity/rb/algo.py](https://app.codecov.io/gh/openvinotoolkit/nncf/blob/develop/nncf/torch/sparsity/rb/algo.py)
- [ ] [nncf/torch/sparsity/rb/loss.py](https://app.codecov.io/gh/openvinotoolkit/nncf/blob/develop/nncf/torch/sparsity/rb/loss.py)

3. Open a PR into [NNCF repo](https://github.com/openvinotoolkit/nncf), wait for the automatic pre-commit checks to pass and for the Codecov to update its automatic status comment with the coverage difference introduced by the PR. Make sure that the Codecov status comment shows an increase of coverage in a given code file (or files) to at least 80% coverage.

### Tips
* Start with the files having the least amount of coverage. 

* Make sure that the test cases you are adding actually execute the lines that are marked as "missed" in the Codecov per-line display (for instance, [nncf/quantization/algorithms/accuracy_control/ranker.py](https://app.codecov.io/gh/openvinotoolkit/nncf/blob/develop/nncf%2Fquantization%2Falgorithms%2Faccuracy_control%2Franker.py)). 

* If at all possible, try to extend existing parametrized tests with proper parameters triggering the missed line logic. 

* Try to add new test cases to the existing test files that contain the tests for the tested component, otherwise - create a new test file in a proper subdirectory with a proper name, use existing test files as a guideline.

* If you encounter dead code (for example, helper functions that aren't used anywhere, only defined) - delete it instead of covering it with tests. This will raise the coverage for the corresponding file better than anything.

* Follow the general coding guidelines in [PyGuide.md](https://github.com/openvinotoolkit/nncf/blob/develop/docs/styleguide/PyGuide.md) when preparing the test code.

**Comment by Candyzorua** [comment]
Hello, I would like to follow-up on this, please :)

**Comment by andrey-churkin** [comment]
Hi @Candyzorua, Sorry for the delay in response. An instance of the `QuantizationAccuracyRestorer` class works with a graph of a quantized model for which weights compression was not applied. The `nncf.quantize()` method compresses weights by default for the OpenVINO backend. Therefore, to resolve the issue, we should turn off weight compression. To do this, please use the following code snippet:

```python
from nncf.openvino.quantization.backend_parameters import BackendParameters

quantized_model = nncf.quantize(
    initial_model,
    dataset,
    advanced_parameters=nncf.AdvancedQuantizationParameters(backend_params={BackendParameters.COMPRESS_WEIGHTS: False})
)
```
Please let me know if it works for you.

**Comment by Candyzorua** [comment]
Hi @andrey-churkin, thank you very much for your help. With your suggestion, I was able to test `_collect_original_biases_and_weights`. 😄  Most functions in `algorithm.py` are now covered, however I am finding it quite difficult to test the longer and more complex `apply` function so I am leaving it out for now.

I have opened a [PR](https://github.com/openvinotoolkit/nncf/pull/2526) with tests for 5 more files regarding quantization and accuracy control functions.

**Comment by RitikaxShakya** [comment]
Hello! @Candyzorua Are all tests/Tasked covered ? If not can i pick some ? @vshampor  I wish to work on this issue :) Thank you.

**Comment by Candyzorua** [comment]
Hi @RitikaxShakya, apologies for the late reply. I haven't covered all the files. Here is a list that shows which files have already been covered. If the checkbox is ticked, it means the file has been covered.

- [x] [nncf/common/utils/decorators.py](https://app.codecov.io/gh/openvinotoolkit/nncf/blob/develop/nncf/common/utils/decorators.py)
- [x] [nncf/common/utils/os.py](https://app.codecov.io/gh/openvinotoolkit/nncf/blob/develop/nncf/common/utils/os.py)
- [x] [nncf/common/utils/patcher.py](https://app.codecov.io/gh/openvinotoolkit/nncf/blob/develop/nncf/common/utils/patcher.py)
- [x] [nncf/common/utils/timer.py](https://app.codecov.io/gh/openvinotoolkit/nncf/blob/develop/nncf/common/utils/timer.py)
- [x] [nncf/config/schema.py](https://app.codecov.io/gh/openvinotoolkit/nncf/blob/develop/nncf/config/schema.py)
- [ ] [nncf/openvino/quantization/quantize_model.py](https://app.codecov.io/gh/openvinotoolkit/nncf/blob/develop/nncf/openvino/quantization/quantize_model.py)
- [x] [nncf/quantization/algorithms/accuracy_control/algorithm.py](https://app.codecov.io/gh/openvinotoolkit/nncf/blob/develop/nncf/quantization/algorithms/accuracy_control/algorithm.py)
- [x] [nncf/quantization/algorithms/accuracy_control/evaluator.py](https://app.codecov.io/gh/openvinotoolkit/nncf/blob/develop/nncf/quantization/algorithms/accuracy_control/evaluator.py)
- [x] [nncf/quantization/algorithms/accuracy_control/openvino_backend.py](https://app.codecov.io/gh/openvinotoolkit/nncf/blob/develop/nncf/quantization/algorithms/accuracy_control/openvino_backend.py)
- [x] [nncf/quantization/algorithms/accuracy_control/ranker.py](https://app.codecov.io/gh/openvinotoolkit/nncf/blob/develop/nncf/quantization/algorithms/accuracy_control/ranker.py)
- [x] [nncf/quantization/algorithms/accuracy_control/subset_selection.py](https://app.codecov.io/gh/openvinotoolkit/nncf/blob/develop/nncf/quantization/algorithms/accuracy_control/subset_selection.py)
- [ ] [nncf/quantization/algorithms/hyperparameter_tuner/algorithm.py](https://app.codecov.io/gh/openvinotoolkit/nncf/blob/develop/nncf/quantization/algorithms/hyperparameter_tuner/algorithm.py)
- [ ] [nncf/quantization/algorithms/hyperparameter_tuner/param_grid.py](https://app.codecov.io/gh/openvinotoolkit/nncf/blob/develop/nncf/quantization/algorithms/hyperparameter_tuner/param_grid.py)
- [ ] [nncf/quantization/algorithms/post_training/algorithm.py](https://app.codecov.io/gh/openvinotoolkit/nncf/blob/develop/nncf/quantization/algorithms/post_training/algorithm.py)
- [ ] [nncf/telemetry/wrapper.py](https://app.codecov.io/gh/openvinotoolkit/nncf/blob/develop/nncf/telemetry/wrapper.py)
- [ ] [nncf/tensorflow/helpers/model_creation.py](https://app.codecov.io/gh/openvinotoolkit/nncf/blob/develop/nncf/tensorflow/helpers/model_creation.py)
- [ ] [nncf/tensorflow/pruning/filter_pruning/functions.py](https://app.codecov.io/gh/openvinotoolkit/nncf/blob/develop/nncf/tensorflow/pruning/filter_pruning/functions.py)
- [ ] [nncf/tensorflow/tensor_statistics/statistics.py](https://app.codecov.io/gh/openvinotoolkit/nncf/blob/develop/nncf/tensorflow/tensor_statistics/statistics.py)
- [ ] [nncf/tensorflow/tf_internals.py](https://app.codecov.io/gh/openvinotoolkit/nncf/blob/develop/nncf/tensorflow/tf_internals.py)
- [ ] [nncf/tensorflow/utils/node.py](https://app.codecov.io/gh/openvinotoolkit/nncf/blob/develop/nncf/tensorflow/utils/node.py)
- [ ] [nncf/tensorflow/utils/scopes_handle.py](https://app.codecov.io/gh/openvinotoolkit/nncf/blob/develop/nncf/tensorflow/utils/scopes_handle.py)
- [ ] [nncf/torch/accuracy_aware_training/utils.py](https://app.codecov.io/gh/openvinotoolkit/nncf/blob/develop/nncf/torch/accuracy_aware_training/utils.py)
- [ ] [nncf/torch/binarization/binarize_functions.py](https://app.codecov.io/gh/openvinotoolkit/nncf/blob/develop/nncf/torch/binarization/binarize_functions.py)
- [ ] [nncf/torch/binarization/extensions.py](https://app.codecov.io/gh/openvinotoolkit/nncf/blob/develop/nncf/torch/binarization/extensions.py)
- [ ] [nncf/torch/pruning/export_utils.py](https://app.codecov.io/gh/openvinotoolkit/nncf/blob/develop/nncf/torch/pruning/export_utils.py)
- [ ] [nncf/torch/pruning/operations.py](https://app.codecov.io/gh/openvinotoolkit/nncf/blob/develop/nncf/torch/pruning/operations.py)
- [ ] [nncf/torch/quantization/base_ctrl.py](https://app.codecov.io/gh/openvinotoolkit/nncf/blob/develop/nncf/torch/quantization/base_ctrl.py)
- [ ] [nncf/torch/quantization/precision_init/autoq_init.py](https://app.codecov.io/gh/openvinotoolkit/nncf/blob/develop/nncf/torch/quantization/precision_init/autoq_init.py)
- [ ] [nncf/torch/quantization/statistics.py](https://app.codecov.io/gh/openvinotoolkit/nncf/blob/develop/nncf/torch/quantization/statistics.py)
- [ ] [nncf/torch/sparsity/const/algo.py](https://app.codecov.io/gh/openvinotoolkit/nncf/blob/develop/nncf/torch/sparsity/const/algo.py)
- [ ] [nncf/torch/sparsity/rb/algo.py](https://app.codecov.io/gh/openvinotoolkit/nncf/blob/develop/nncf/torch/sparsity/rb/algo.py)
- [ ] [nncf/torch/sparsity/rb/loss.py](https://app.codecov.io/gh/openvinotoolkit/nncf/blob/develop/nncf/torch/sparsity/rb/loss.py)

I think it's a great idea to split up the work. Perhaps you can work on the remaining quantization-related files or the TensorFlow ones? I am continuing work on the PyTorch ones for now. Please let me know, thank you!

**Comment by RitikaxShakya** [comment]
@Candyzorua Hello! Thank you for replying :), That's great! i was already looking for to pick up TensorFlow and quantization related files, so i will pick up working on these files! Thank you!

**Comment by DaniAffCH** [comment]
Hi, so if @RitikaxShakya takes Tensorflow and Quantized I'll cover the remaining ones

**Comment by Candyzorua** [comment]
Hi @DaniAffCH , since I have started on nncf/torch/quantization, I think it would be awesome if you could work on nncf/torch/sparsity, nncf/torch/pruning, as well as nncf/telemetry. I think nncf/torch/binarization has been removed. Please let me know what you think, thanks!

**Comment by DaniAffCH** [comment]
@Candyzorua thanks for letting me know! Yes it's ok, I'll cover the unassigned files :)

**Comment by anzr299** [comment]
Hi, what's the status of the files covered? I would like to contribute as well.

**Comment by RitikaxShakya** [comment]
Hello! I've almost covered up TensorFlow related files, and finishing on Quantization files.

**Comment by Candyzorua** [comment]
Hi @anzr299 , sorry for the late reply. Please feel free to take over nncf/torch/quantization! I will be stopping work on this issue right now due to work responsibilities. Thank you!

**Comment by anzr299** [comment]
> Hi @anzr299 , sorry for the late reply. Please feel free to take over nncf/torch/quantization! I will be stopping work on this issue right now due to work responsibilities. Thank you!

Sure, no problem. I will work on it.

**Comment by zina-cs** [comment]
Hello, @anzr299, @vshampor, @DaniAffCH! I’d like to contribute to this issue too, could you let me know your assigned files so I work on the rest?

**Comment by alexsu52** [comment]
Hello @zina-cs, 
Thanks for your interest! You can message ".take" command to assign issue on yourself.

**Comment by alexsu52** [comment]
Hello @zina-cs,

I closed the issue because GA for code coverage measurement was disabled in the NNCF repository and we do not have working tools for code coverage measurement. Please choose another issue. Sorry for the inconvenience.
